tests/TestCase/Authenticator/PasetoAuthenticatorTest.php

```diff
@@ -0,0 +1,553 @@
+            'purpose' => PasetoAuthenticator::LOCAL,
+            'version' => $version,
+        ]);
+
+        $result = $authenticator->authenticate($request);
+        $this->assertInstanceOf(Result::class, $result);
+        $this->assertSame(Result::SUCCESS, $result->getStatus());
+        $this->assertInstanceOf(ArrayAccess::class, $result->getData());
+        $this->assertIsArray($result->getData()['footer']);
+        $this->assertEquals('larry', $result->getData()['username']);
+    }
+
+    /**
+     * Test authentication with local purpose via header token.
+     *
+     * @dataProvider dataProviderForVersions
+     * @param string $version The PASETO version
+     * @param ProtocolInterface $protocol Instance of Version
+     * @return void
+     * @throws \ParagonIE\Paseto\Exception\InvalidKeyException
+     * @throws \ParagonIE\Paseto\Exception\InvalidPurposeException
+     * @throws \ParagonIE\Paseto\Exception\PasetoException
+     */
+    public function testPublicAuthenticationViaHeaderToken(string $version, ProtocolInterface $protocol): void
+    {
+        $request = ServerRequestFactory::fromGlobals(
+            ['REQUEST_URI' => '/']
+        );
+        /** @var Builder $token extracted */
+        /** @var AsymmetricSecretKey $privateKey extracted */
+        extract($this->buildPublicToken($protocol));
+        $request = $request->withAddedHeader('Authorization', 'Bearer ' . $token->toString());
+
+        $authenticator = new PasetoAuthenticator($this->identifiers, [
+            'secretKey' => $privateKey->encode(),
+            'purpose' => PasetoAuthenticator::PUBLIC,
+            'version' => $version,
+        ]);
+
+        $result = $authenticator->authenticate($request);
+        $this->assertInstanceOf(Result::class, $result);
+        $this->assertSame(Result::SUCCESS, $result->getStatus());
+        $this->assertInstanceOf(ArrayAccess::class, $result->getData());
+        $this->assertIsArray($result->getData()['footer']);
+        $this->assertEquals('larry', $result->getData()['username']);
+    }
+
+    /**
+     * Test authentication via query parameter.
+     *
+     * @dataProvider dataProviderForPurpose
+     * @param string $purpose Either local or public
+     * @return void
+     * @throws \ParagonIE\Paseto\Exception\InvalidKeyException
+     * @throws \ParagonIE\Paseto\Exception\InvalidPurposeException
+     * @throws \ParagonIE\Paseto\Exception\PasetoException
+     */
+    public function testAuthenticationViaQueryParamToken(string $purpose): void
+    {
+        switch ($purpose) {
+            case PasetoAuthenticator::LOCAL:
+                $token = $this->buildLocalToken(new Version4());
+                break;
+            case PasetoAuthenticator::PUBLIC:
+                /** @var Builder $token extracted */
+                /** @var AsymmetricSecretKey $privateKey extracted */
+                extract($this->buildPublicToken(new Version4()));
```

Comment:
This feels risky to me. Having a method arbitrarily update local scope has not worked out well for us in the past. Can we be more explicit with what is being set in local scope here.